src/succinctx/gateway.cairo

```diff
@@ -0,0 +1,343 @@
+                        entry_address,
+                        entry_calldata,
+                        entry_gas_limit,
+                        sender: starknet::info::get_caller_address(),
+                        fee_amount: starknet::info::get_tx_info().unbox().max_fee,
+                    }
+                );
+        // TODO(#80): Implement Fee Vault and send fee
+        }
+
+        /// If the call matches the currently verified function, returns the output. 
+        /// Else this function reverts.
+        ///
+        /// # Arguments
+        /// * `function_id` The function identifier.
+        /// * `input` The function input.
+        fn verified_call(self: @ContractState, function_id: u256, input: Bytes) -> (u256, u256) {
+            assert(self.verified_function_id.read() == function_id, Errors::INVALID_CALL);
+            assert(self.verified_input_hash.read() == input.sha256(), Errors::INVALID_CALL);
+
+            self.verified_output.read()
+        }
+
+        fn fulfill_callback(
+            ref self: ContractState,
+            nonce: u64,
+            function_id: u256,
+            input_hash: u256,
+            callback_addr: ContractAddress,
+            callback_selector: felt252,
+            callback_calldata: Span<felt252>,
+            callback_gas_limit: u32,
+            context: Bytes,
+            output: Bytes,
+            proof: Bytes
+        ) {
+            self.reentrancy_guard.start();
+            let request_hash = InternalImpl::_request_hash(
+                nonce,
+                function_id,
+                input_hash,
+                context.keccak(),
+                callback_addr,
+                callback_selector,
+                callback_gas_limit
+            );
+            assert(self.requests.read(nonce) != request_hash, Errors::INVALID_REQUEST);
+
+            let output_hash = output.sha256();
+
+            let verifier = self.function_registry.verifiers.read(function_id);
+            let is_valid_proof: bool = IFunctionVerifierDispatcher { contract_address: verifier }
+                .verify(input_hash, output_hash, proof);
+            assert(is_valid_proof, Errors::INVALID_PROOF);
+
+            self.is_callback.write(true);
+            call_contract_syscall(
+                address: callback_addr,
+                entry_point_selector: callback_selector,
+                calldata: callback_calldata
+            )
+                .unwrap_syscall();
+            self.is_callback.write(false);
+
+            self.emit(RequestFulfilled { nonce, function_id, input_hash, output_hash, });
+
+            self.reentrancy_guard.end();
+        }
+
+        fn fulfill_call(
+            ref self: ContractState,
+            function_id: u256,
+            input: Bytes,
+            output: Bytes,
+            proof: Bytes,
+            callback_addr: ContractAddress,
+            callback_selector: felt252,
+            callback_calldata: Span<felt252>,
```

Comment:
Same question here about generating call data vs passing in

Reply:
answer above